File: Python_Solutions/Knapsack/Zero_One_Knapsack/Zero_One_Knapsack_DP_memoization.py
# Given constraints W<= 1002 and n<=102
def knapSackDP( wt, val, W, n):
    t = [[-1 for j in range(W+1)] for i in range(n+1)]
    if (n==0 or W ==0):
        return 0
    elif (t[n][W] != -1):
        return t[n][W]
    elif (wt[n-1]<=W):
        t[n][W] = max(
            val[n-1] + knapSackDP(wt,val,W-wt[n-1],n-1),
            knapSackDP(wt,val,W,n-1)
        )
    else:
        t[n][W] = knapSackDP(wt,val,W,n-1)
    return t[n][W]

val = [60, 100, 120]
wt = [10, 20, 30]
W = 50
n = len(val)
print(knapSackDP(wt, val, W, 3))

File: CPP_Solutions/Knapsack/Zero_One_Knapsack/Zero_One_Knapsack_DP_memoization.cpp
#include <iostream>
#include <cstring>

using namespace std;

// Considering Constraints n<=100 and W<=1000

int t[102][1002];

int knapsackDP(int wt[], int val[], int W, int n) {
    if(n == 0 || W == 0) // Base Condition
        return 0;
    else if (t[n][W] != -1) // If any value apart from -1 is present at the last(upper limit) position
        return t[n][W];
    else if(wt[n-1] <= W){
        return t[n][W] = std::max(val[n-1] + knapsackDP(wt, val, W - wt[n-1], n-1),knapsackDP(wt, val, W, n-1));
    }
    else { //wt[n-1] > W
        return t[n][W] = knapsackDP(wt, val, W, n-1);
    }
}

int main()
{
    // cout<<"Hello World";
    memset(t, -1, sizeof(t));
    int weight[] = { 10, 20, 30 };
    int val[] = { 60, 100, 120 };
    int W = 50;
    int n = sizeof(val) / sizeof(val[0]);
    cout << knapsackDP(weight, val, W, n);

    return 0;
}